### DISCUSSION TITLE
Support symfony/validation attributes and annotations

### DISCUSSION THREAD

**Discussion by tigitz** [morrislaptop/popo-factory#4] [Ideas]
Would be useful for DTOs to support `symfony/validation` attributes and annotations like

```php
namespace App\Entity;

use Symfony\Component\Validator\Constraints as Assert;

class Author
{
    #[Assert\Choice(['New York', 'Berlin', 'Tokyo'])]
    protected $city;
}
``` 

Would produce something like `new Author($city = Faker::randomElement(['New York', 'Berlin', 'Tokyo'])`

**Comment by morrislaptop** [comment]
I'm wary of adding Symfony specific things into the library.

This might be better as a utility trait that could be made available to class based factories.. Something like...

```php
class PersonDataFactory extends PopoFactory
{
    use SymfonyUtils;

    public static function factory(): static
    {
        return static::new(PersonData::class)->state(
            $this->symfonyValidatorState()
        ]);
    }
```

Or alternatively we could inject the PropertyFactory into the PopoFactory. You could then choose to use a SymfonyValidatorPropertyFactory which could take the annotations into account.

**Reply by tigitz** [reply]
Indeed it shouldn't be a hard dependency and I would go with the alternative you suggest.

The way it can be designed is to use dependency injection to make the PopoFactory aware of its `symfony/validator` support abilities.

For exemple, if you want to support php types, php doc block and symfony/validator in the `PopoFactory` logic it could be injected as such.
```php
$popoFactory = new PopoFactory([new PhpTypeParser(), new PhpDocBlockParser(), new SymfonyValidator()])

$person = $popoFactory->fake(PersonData::class);
```

That would introduce a `SymfonyValidator` class in the lib referencing `symfony/validator` classes but that's ok, it won't error if your project doesnt require `symfony/validator` as long as you create your factory without the `symfony/validator` support like this `new PopoFactory([new PhpTypeParser(), new PhpDocBlockParser()])`

**Comment by morrislaptop** [comment]
Love this approach! Very similar to the Symfony Serializer as well
